app/code/Magento/Eav/Setup/AddOptionToAttribute.php

```diff
@@ -0,0 +1,202 @@
+            $this->addValues($option, $optionTable, $optionValueTable);
+        }
+    }
+
+    /**
+     * @param array  $option
+     * @param string $optionTable
+     * @param string $optionValueTable
+     *
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
+    private function addValue(array $option, string $optionTable, string $optionValueTable)
+    {
+        $value = $option['value'];
+        foreach ($value as $optionId => $values) {
+            $intOptionId = (int)$optionId;
+            if (!empty($option['delete'][$optionId])) {
+                if ($intOptionId) {
+                    $condition = ['option_id =?' => $intOptionId];
+                    $this->setup->getConnection()->delete($optionTable, $condition);
+                }
+                continue;
+            }
+
+            if (!$intOptionId) {
+                $data = [
+                    'attribute_id' => $option['attribute_id'],
+                    'sort_order'   => isset($option['order'][$optionId]) ? $option['order'][$optionId] : 0,
+                ];
+                $this->setup->getConnection()->insert($optionTable, $data);
+                $intOptionId = $this->setup->getConnection()->lastInsertId($optionTable);
+            } else {
+                $data = [
+                    'sort_order' => isset($option['order'][$optionId]) ? $option['order'][$optionId] : 0,
+                ];
+                $this->setup->getConnection()->update(
+                    $optionTable,
+                    $data,
+                    ['option_id=?' => $intOptionId]
+                );
+            }
+
+            // Default value
+            if (!isset($values[0])) {
+                throw new \Magento\Framework\Exception\LocalizedException(
+                    __("The default option isn't defined. Set the option and try again.")
+                );
+            }
+            $condition = ['option_id =?' => $intOptionId];
+            $this->setup->getConnection()->delete($optionValueTable, $condition);
+            foreach ($values as $storeId => $value) {
+                $data = ['option_id' => $intOptionId, 'store_id' => $storeId, 'value' => $value];
+                $this->setup->getConnection()->insert($optionValueTable, $data);
+            }
+        }
+    }
+
+    /**
+     * @param array  $option
+     * @param string $optionTable
+     * @param string $optionValueTable
+     */
+    private function addValues(array $option, string $optionTable, string $optionValueTable)
+    {
+        $values = $option['values'];
+        $attributeId = (int)$option['attribute_id'];
+        $existingOptions = $this->getExistingAttributeOptions($attributeId, $optionTable, $optionValueTable);
+        foreach ($values as $sortOrder => $value) {
+            // add option
+            $data = ['attribute_id' => $attributeId, 'sort_order' => $sortOrder];
+            if (!$this->isExistingOptionValue($value, $existingOptions)) {
```

Comment:
@novikor this is not a right way to validate uniqueness.

Unique index must be added on database level instead and then constraint violation properly handled on PHP side.

Reply:
And what could you suggest instead?

Reply:
@orlangur, I have no time to rebuild my solution. Moreover, it is successfully covered by the tests already.

You are right this should be handled on the database level - however, it is complex enough and should be analyzed.
I suggest to merge this PR and create a new issue to improve it.

**Moreover, adding new unique index could cause backward incompatibility.**

Reply:
@novikor,

> adding new unique index could cause backward incompatibility.

how? if the data is invalid, there will be an error during upgrade.

Reply:
Ok, from https://github.com/magento/magento2/pull/20852#issuecomment-484011399 I figured out that simple solution on database layer is not possible.